### ISSUE TITLE
Problem opening different files with same filename in different layouts

### ISSUE THREAD

**Issue by dpfrey** [syl20bnr/spacemacs#10183]
#### Description :octocat:
Buffer switched to scratch in one layout when opening a different file with the same filename in another layout

#### Reproduction guide
- `$ mkdir -p ~/bug/a`
- `$ mkdir -p ~/bug/b`
- `$ echo foo > ~/bug/a/foo.txt`
- `$ echo FOO > ~/bug/b/foo.txt`
- Start Emacs
- Use [SPC l 2] and create a layer named "a"
- Open *~/bug/a/foo.txt* in layout *a*
- Use [SPC l 3] and create a layer named "b"
- Open *~/bug/b/foo.txt* in layout *b*
- Switch back to layout *a*

*Observed behaviour:*
The scratch buffer is visible in layout *a*

*Expected behaviour:*
*~/bug/a/foo.txt* should be visible in layout *a*

#### System Info :computer:
- OS: gnu/linux
- Emacs: 25.1.1
- Spacemacs: 0.200.10
- Spacemacs branch: master (rev. 4bb4cb46)
- Graphic display: nil
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(javascript git cscope emacs-lisp c-c++ python yaml markdown)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GCONF GSETTINGS NOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11

**Comment by ynilu** [comment]
I can confirm this. It's very annoying to me.

**Comment by duianto** [comment]
Confirmed on both the master and develop (c130014f not the latest commit) branches.

#### Observations:
After having created layout `3:b` and opened `../b/foo.txt`.
Then the global-buffers list `SPC B b` (on the master branch), `SPC b B` (on develop) shows:
both `foo.txt` files as:
- `foo.txt<a>`
- `foo.txt<b>`

(`<a>` and `<b>` are the directory names.)

So they should already have unique names. But after the last step:
>Switch back to layout a

The minibuffer shows:
>Replaced deleted foo.txt buffer with *scratch*

`SPC b b` shows that `foo.txt<a>` is still open in layout `a`, and from this point on, switching between the layouts, works without the `*scratch*` buffer replacing the previously opened file.

If one saves the layout `SPC l s` after creating the `3:b` layout and opening `../b/foo.txt`. Then the issue can be replicated, by loading the layout and switching to layout `2:a` `SPC l 2`.

If the layout is saved after one has switched to layout `2:a`, and the issue has occurred, then one can switch layouts without any issues.

#### Possible cause:
I saved two layouts, one before and one after switching to layout 2, and compared them.

The buffer name in the before layout, doesn't have `<a>` after the filename.

Before:
```
(buffer "foo.txt" (selected . t) (hscroll . 0) (fringes 8 8 nil) (margins nil) (scroll-bars nil 0 t nil 0 t) (vscroll . 0) (dedicated) (point . 6) (start . 1))
```

After:
```
(buffer "foo.txt<a>" (selected . t) (hscroll . 0) (fringes 8 8 nil) (margins nil) (scroll-bars nil 0 t nil 0 t) (vscroll . 0) (dedicated) (point . 6) (start . 1))
```

#### Possible solution?
Maybe the buffer name in the layout, also needs to be updated, when a file is opened, and there already is a file with the same name open.

#### System Info :computer:
- OS: windows-nt
- Emacs: 25.3.1
- Spacemacs: 0.200.13
- Spacemacs branch: master (rev. 582f9aa4)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(autohotkey helm emacs-lisp git org spell-checking themes-megapack treemacs version-control)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG SOUND NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS MODULES

and on the develop branch:
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. c130014f)

**Comment by Compro-Prasad** [comment]
Creating buffers and isolating them are two very different things.

The buffer names are given by Emacs itself and they also act as an UID. You can look into `uniquify-buffer-name-style` for different types of buffer unification mechanisms.

What can I suggest:
I can suggest that layouts can use `rename-buffer` to make this more layout friendly. But it will be bit of work on the layout side as it will have to check if similar buffer names exist or not in the same layout and will have to conditionally and differently rename them.

**Comment by dpfrey** [comment]
The message "Replaced deleted foo.txt buffer with scratch" is generated by the function `eyebrowse--fixup-window-config` which I believe is being called by `eyebrowse--load-window-config`.  The more interesting thing seems to be the `defadvice rename-buffer` definition in the `eyebrowse.el`.  This seems to attempt to intercept `rename-buffer` calls and then use `eyebrowse--rename-window-config-buffers` to update the buffer names in the window config.  It doesn't seem to work quite the way it should.  I don't really know elisp, so I'm having trouble digging in any deeper.  Hopefully this is enough of a clue for someone.

**Comment by Miciah** [comment]
@dpfrey, #11010 should fix the problem.  Would you mind testing it?

**Comment by dpfrey** [comment]
Thanks so much @Miciah for working on this!  The original test case that I provided passes, but I found a slightly more complicated example that still behaves badly.  Here are the steps:

- `$ mkdir -p ~/bug/a/src`
- `$ mkdir -p ~/bug/b/src`
- `$ echo foo > ~/bug/a/src/foo.txt`
- `$ echo FOO > ~/bug/b/src/foo.txt`
- Start Emacs
- Use [SPC l 2] and create a layer named "a"
- Enter `:e ~/bug/a/src`
- VSplit the window using `ctrl-w-v`
- Press enter on `foo.txt` to open it in one of the splits
-  Use [SPC l 3] and create a layer named "b"
- Enter `:e ~/bug/b/src`
- VSplit the window using `ctrl-w-v`
- Press enter on `foo.txt` to open it in one of the splits
- Switch back to layout *a*

What I see in this case is that the dired split from layout "a" is the scratch buffer.  The foo.txt buffer in layout "a" has successfully been renamed to "foo.txt<a>" which is an improvement compared to the previous behavior.

Is there a difference between regular file buffers and dired buffers that would account for the difference in behavior?

**Comment by Miciah** [comment]
You're welcome, and thanks for the test cases! At first I could not reproduce the problem using your newest test case and my custom `.spacemacs`, but I could reproduce it using a clean install of Spacemacs.  Eventually I determined that the following configuration is needed together with the changes in #11010:
```emacs-lisp
(defun dotspacemacs/layers ()
  ;; ...
  (setq-default
     ;; ...
     (spacemacs-layouts :variables
                        persp-add-buffer-on-after-change-major-mode 'free)))
```
I believe the issue exposed by your latest test case is essentially the same issue as #5776.  Do setting `persp-add-buffer-on-after-change-major-mode` to `'free` and applying the fix in #11010 solve the problem for you?

**Comment by dpfrey** [comment]
I tried adding the config fragment you mentioned and I got this:

`Error in dotspacemacs/layers: Wrong type argument: symbolp, (spacemacs-layouts :variables persp-add-buffer-on-after-change-major-mode (quote free))`

**Comment by Miciah** [comment]
Probably because I omitted too much context:
```emacs-lisp
(defun dotspacemacs/layers ()
  ;; ...
  (setq-default
   ;; ...
   ;; List of configuration layers to load.
   dotspacemacs-configuration-layers
   '(
      ;; ...
      (spacemacs-layouts :variables
                         persp-add-buffer-on-after-change-major-mode 'free)
      ;; ...
    )
    ;; ...
  ))
```

That is, the `(spacemacs-layouts :variables persp-add-buffer-on-after-change-major-mode 'free)` is supposed to be inside the existing `dotspacemacs-configuration-layers '()` list in `.spacemacs`.  Alternatively, to change the setting in your current session, you should be able to use `M-:` `(setq persp-add-buffer-on-after-change-major-mode 'free)`.

Or maybe you figured that out despite my previous misleading comment, in which case something very strange is going on.

**Comment by dpfrey** [comment]
@Miciah: Your fix in combination with the configuration fragment seems to solve the second test case as well.  It seems like a bit of a workaround rather than a fix if you have to add extra config to get sensible behavior though.  Thanks!

**Comment by Miciah** [comment]
Agreed that setting `persp-add-buffer-on-after-change-major-mode` is a workaround.  Your test cases reveal two distinct problems:

* Your first test case illustrates that `eyebrowse--fixup-window-config ` fails to rename buffers in window configurations when those window configurations are stored in perspective parameters (which is how Spacemacs integrates eyebrowse with persp-mode).

* Your second test case/#5776 illustrates that my `spacemacs//fixup-window-configs` function fails to rename buffers in window configurations when those buffers are not associated with any perspective.

#5776 exacerbates the latter issue because buffers that should be associated with a perspective are not.  This is what setting `persp-add-buffer-on-after-change-major-mode` is working around.  (Perhaps a proper fix for #5776 would be simply to make that setting the default, or perhaps something more clever is needed.) However, fixing #5776 would still leave the edge case of buffers temporarily added to a perspective using `SPC l t` (`persp-temporarily-display-buffer`).

My `spacemacs//fixup-window-configs` function uses `persp--buffer-in-persps` to find the perspectives with which the buffer is associated, and then it fixes up the window configurations stored in those perspectives' parameters; perhaps my function should simply iterate over all perspectives, irrespective of whether the buffer is associated.  That seems a little heavy-handed but may be the only way to be sure to find all references to the buffer.

@bmag, do you have any suggestions?

**Comment by syl20bnr** [comment]
Let set `persp-add-buffer-on-after-change-major-mode-filter-functions` to `free` in spacemacs default config. It seems to me to be a better default, not ideal but better than current https://github.com/syl20bnr/spacemacs/issues/5776 behavior.

**Comment by Miciah** [comment]
I made #11125 to fix #5776.  If that PR gets merged, I hope #11010 can be next.
